easybuild/tools/module_naming_scheme/categorized_hmns.py

```diff
@@ -0,0 +1,106 @@
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+##
+"""
+Implementation of a hierarchical module naming scheme using module classes.
+
+@author: Markus Geimer (Forschungszentrum Juelich GmbH)
+"""
+
+import os
+
+from easybuild.tools.module_naming_scheme.hierarchical_mns import HierarchicalMNS
+from easybuild.tools.config import build_option
+
+
+class CategorizedHMNS(HierarchicalMNS):
+    """
+    Class implementing an extended hierarchical module naming scheme using the
+    'moduleclass' easyconfig parameter to categorize modulefiles on each level
+    of the hierarchy.
+    """
+
+    REQUIRED_KEYS = ['name', 'version', 'versionsuffix', 'toolchain', 'moduleclass']
+
+    def det_module_subdir(self, ec):
+        """
+        Determine module subdirectory, relative to the top of the module path.
+        This determines the separation between module names exposed to users,
+        and what's part of the $MODULEPATH. This implementation appends the
+        'moduleclass' easyconfig parameter to the base path of the corresponding
+        hierarchy level.
+
+        Examples:
+        Core/compiler, Compiler/GCC/4.8.3/mpi, MPI/GCC/4.8.3/OpenMPI/1.6.5/bio
+        """
+        moduleclass = ec['moduleclass']
+        basedir = super(CategorizedHMNS, self).det_module_subdir(ec)
+
+        return os.path.join(basedir, moduleclass)
+
+    def det_modpath_extensions(self, ec):
+        """
+        Determine module path extensions, if any. Appends all known (valid)
+        module classes to the base path of the corresponding hierarchy level.
+
+        Examples:
+        Compiler/GCC/4.8.3/<moduleclasses> (for GCC/4.8.3 module),
+        MPI/GCC/4.8.3/OpenMPI/1.6.5/<moduleclasses> (for OpenMPI/1.6.5 module)
+        """
+        known_module_classes = build_option('valid_module_classes')
+        basepaths = super(CategorizedHMNS, self).det_modpath_extensions(ec)
+
+        paths = []
+        for path in basepaths:
+            for moduleclass in known_module_classes:
+                paths.extend([os.path.join(path, moduleclass)])
+
+        return paths
+
+    def det_init_modulepaths(self, ec):
+        """
+        Determine list of initial module paths (i.e., top of the hierarchy).
+        Appends all known (valid) module classes to the top-level base path.
+
+        Examples:
+        Core/<moduleclasses>
+        """
+        known_module_classes = build_option('valid_module_classes')
+        basepaths = super(CategorizedHMNS, self).det_init_modulepaths(ec)
+
+        paths = []
+        for path in basepaths:
+            for moduleclass in known_module_classes:
+                paths.extend([os.path.join(path, moduleclass)])
```

Comment:
this is a pretty clean list comprehension, we have many others... ;-)

but sure, go ahead and do it via a separate method then, I don't feel strongly about making it a list comprehension, since I won't have to maintain this bit ;)